rpc/handlers.go

```diff
@@ -1120,3 +1103,71 @@ func (h *Handler) EstimateFee(broadcastedTxns []BroadcastedTransaction, id Block
 
 	return estimates, nil
 }
+
+// TraceTransaction returns the trace for a given executed transaction, including internal calls
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/1ae810e0137cc5d175ace4554892a4f43052be56/api/starknet_trace_api_openrpc.json#L11
+func (h *Handler) TraceTransaction(hash felt.Felt) (TransactionTrace, *jsonrpc.Error) {
+	_, _, blockNumber, err := h.bcReader.Receipt(&hash)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+
+	state, closer, err := h.bcReader.StateAtBlockNumber(blockNumber - 1)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer h.callAndLogErr(closer, "Failed to close state in starknet_traceTransaction")
+
+	block, err := h.bcReader.BlockByNumber(blockNumber)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	headState, headStateCloser, err := h.bcReader.HeadState()
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+	defer h.callAndLogErr(headStateCloser, "Failed to close head state in starknet_traceTransaction")
+
+	var (
+		txIndex int
+		tx      core.Transaction
+	)
+	var classes []core.Class
+	for i, transaction := range block.Transactions {
+		if transaction.Hash().Equal(&hash) {
+			txIndex = i
+			tx = transaction
+		}
+
+		declareTx, ok := transaction.(*core.DeclareTransaction)
+		if !ok {
+			continue
+		}
+
+		class, stateErr := headState.Class(declareTx.ClassHash)
+		if stateErr != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, stateErr.Error())
+		}
+		classes = append(classes, class.Class)
+	}
+
+	header := block.Header
+	_, info, err := h.vm.Execute(block.Transactions, classes, header.Number, header.Timestamp, header.SequencerAddress, state, h.network)
```

Comment:
We need to check for pending block explicitly, since header.Number is not valid for pending blocks. See how we do that in `EstimateFee`

Reply:
Added support